File: projects/01-smart-agriculture/backend/processor/handler.py
import json
import os
from decimal import Decimal

import boto3

from process import process

TABLE_NAME = os.getenv("TABLE_NAME", "fec-readings")
ENDPOINT = os.getenv("AWS_ENDPOINT_URL")
REGION = os.getenv("AWS_REGION", "eu-west-1")

_table = boto3.resource("dynamodb", endpoint_url=ENDPOINT, region_name=REGION).Table(TABLE_NAME)


def to_item(record):
    # DynamoDB's resource-level Table API rejects native Python float; round
    # tripping through json with parse_float=Decimal converts every float in
    # the record into a Decimal without hand-walking the dict.
    return json.loads(json.dumps(record), parse_float=Decimal)


def lambda_handler(event, context):
    """SQS-triggered entry point, wired up by an event source mapping: one invocation carries
    a batch of queue messages in event["Records"], one per window-aggregate
    the fog node published."""
    for record in event["Records"]:
        _table.put_item(Item=to_item(process(record["body"])))
    return {"processed": len(event["Records"])}

File: projects/18-elevator-escalator-fleet-monitoring/backend/processor/handler.js
"use strict";

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand } = require("@aws-sdk/lib-dynamodb");
const { toItem } = require("./transform");

const TABLE_NAME = process.env.TABLE_NAME || "eef-readings";

let client;
function documentClient() {
  if (client) return client;
  // Static creds only when AWS_ENDPOINT_URL (LocalStack) is set; otherwise defer to the SDK provider chain.
  const config = { region: process.env.AWS_REGION || "eu-west-1" };
  if (process.env.AWS_ENDPOINT_URL) {
    config.endpoint = process.env.AWS_ENDPOINT_URL;
    config.credentials = { accessKeyId: "test", secretAccessKey: "test" };
  }
  client = DynamoDBDocumentClient.from(new DynamoDBClient(config));
  return client;
}

async function writeBatch(records, doc, tableName) {
  let written = 0;
  for (const record of records) {
    await doc.send(new PutCommand({ TableName: tableName, Item: toItem(record.body) }));
    written += 1;
  }
  return written;
}

// SQS event-source-mapping invokes this per batch; each record is one window aggregate.
exports.handler = async (event) => {
  const written = await writeBatch(event.Records || [], documentClient(), TABLE_NAME);
  return { written };
};

exports.writeBatch = writeBatch;
